### PR TITLE
Allow to pass a `kernel` argument to `RawKernel`

### PR DESCRIPTION
**Pull request by emcastillo**
Similar approach to #2784 
prevents to instantiate a `RawKernel` object without code, or an actual kernel obtained from the compilation backend.

Thanks to @niboshi

### CODE REVIEW

**Comment by leofang** [comment]
Looks nice. I have two questions:
1. > or an actual kernel obtained from the compilation backend.

what does this mean?

2. I thought `cupy.cuda.function.Function` is a private API (which was presumably why we had `RawKernel` on top?). Is it necessary to document the `kernel` argument?

**Comment by emcastillo** [comment]
You are right, should we leave the arguments docstring untouched and describe this in a separate comment?

**Comment by niboshi** [comment]
About 2., I thought it's a public class (there's a shorter alias `cupy.cuda.Function`), but I'm not so sure. It's not documented.

**Comment by emcastillo** [comment]
@leofang I changed the description of the `kernel` arg, since it seems to be public let me know what you think

**Comment by niboshi** [comment]
With `kernel` given, all other arguments except `enable_cooperative_groups` are meaningless as well.
I now feel it's not optimal in tems of modularity to add `kernel` argument to `RawKernel`.

**Comment by emcastillo** [comment]
I see the point, an approach to tackle this could be to have some factory method for compiling the raw kernel itself while the actual kernel class just takes care of execution parameters.
Right now  `RawKernel` is in charge of calling the compiler for the code and the actual execution. These two responsibilities could be separated.

**Comment by niboshi** [comment]
>I see the point, an approach to tackle this could be to have some factory method for compiling the raw kernel itself while the actual kernel class just takes care of execution parameters.

I'm afraid that's not possible without breaking compatibility because `RawKernel.__init__` already has arguments for compilation.
(IMO `__init__` exactly defines the role of a class in general. In this point of view, `RawKernel` has been defined to exist for compilation (and additionally, for execution)).
 
I feel it's worth exploring the possibility to separate the roles.

**Comment by niboshi** [comment]
Maybe `cupy.cuda.Function` could be the class for execution and `RawModule.get_function` could simply return `Function`?
`RawKernel`'s properties such as `max_threads_per_block` could be implemented in `Function` as well.

**Comment by emcastillo** [comment]
> Maybe cupy.cuda.Function could be the class for execution and RawModule.get_function could simply return Function?
RawKernel's properties such as max_threads_per_block could be implemented in Function as well.

This sounds a good behavior that maybe `ElementwiseKernel` and `ReductionKernel` could benefit from as they use the `cuda.Function` to hold the actual kernels.

**Comment by leofang** [comment]
> an approach to tackle this could be to have some factory method for compiling the raw kernel itself while the actual kernel class just takes care of execution parameters.

I think `_get_raw_kernel()` was meant to be the factory method?

> Maybe `cupy.cuda.Function` could be the class for execution and `RawModule.get_function` could simply return `Function`?
> `RawKernel`'s properties such as `max_threads_per_block` could be implemented in `Function` as well.

This is interesting. As long as the following semantics is maintained:
```python
mod = cp.RawModule(...)
ker = mod.get_function('the_kernel')
ker. max_dynamic_shared_size_bytes = 1024  # set size
ker(a)  # run the kernel with array a
```
I think the backward compatibility of `RawModule` will be intact. Users won't notice `ker` becomes a `cuda.Function` instead of a `RawKernel`.

But, one small drawback I see for separating the roles is that `RawKernel` still needs to have all the properties it currently has (which will wrap on top of `cuda.Function`'s counterparts). In addition, `RawKernel.attributes` should continue to return a dict, and we don't need `cuda.Function.attributes`.

**Comment by niboshi** [comment]
>But, one small drawback I see for separating the roles is that `RawKernel` still needs to have all the properties it currently has (which will wrap on top of `cuda.Function`'s counterparts).

I feel it's natural to keep such features in`RawKernel` (as a convenient addition to the "main" role).

>In addition, `RawKernel.attributes` should continue to return a dict, and we don't need `cuda.Function.attributes`.

I don't see why we don't need `Function.attributes`.
If users could be using it via `RawKernel`, should `Function` also need it for their existing code to work?

e.g.
```py
mod = cp.RawModule(...)
ker = mod.get_function('the_kernel')
ker.attributes  # Would need `Function.attributes`
```

**Comment by leofang** [comment]
> If users could be using it via `RawKernel`, should `Function` also need it for their existing code to work?

@niboshi I guess you're right. I was thinking the current design, sorry. 

But I still think there would be some code repetition for the properties. What if we implement a `@classmethod` in `RawKernel` that takes a `cuda.Function` instance and warps it as a `RawKernel`? This would not only circumvent the problem of setting additional argument in the constructor, but also avoid both the splitting effort and code repetition.

**Comment by niboshi** [comment]
>But I still think there would be some code repetition for the kernel properties. What if we implement a `@classmethod` in `RawKernel` that takes a `cuda.Function` instance and returns a `RawKernel`? This would not only circumvent the problem of adding additional argument to the constructor, but also avoid both the splitting effort and code repetition.

The classmethod would have to call `__init__` anyway, with all the arguments ignored. This is not clean.

Separation of roles is the most important point here IMO.
I'd rather move all the properties to `Function` to avoid repetition and keep responsibilities clear (but that would break compatibility of course).

**Comment by leofang** [comment]
> The classmethod would have to call `__init__` anyway, with all the arguments ignored. This is not clean.

Putting the goal to accomplish (initiating a `RawKernel` with a `cuda.Function` instance) aside, which I am not sure why it is necessary to be a public API, this PR basically asks for an alternative constructor. From this point of view, refactoring the current implementation is just an overkill, which I think would make the abstraction less clean than it currently is. 

For alternative/overloaded constructors (or factory functions), Cython provides an example for how to do this: https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#instantiation-from-existing-c-c-pointers, which I followed in `cupy.cuda.nccl.NcclCommunicator`. There is no need to call `__init__`, just `__new__`: https://github.com/cupy/cupy/blob/a9e5ff3cb0e25891e0588b7206ec04cf9a78dac0/cupy/cuda/nccl.pyx#L306-L311 One minor thing to note, though, is that `@classmethod` in Cython, if I recall correctly, was not implemented correctly and led to compile time errors, which was why I used `@staticmethod` instead. 

> Separation of roles is the most important point here IMO.
> I'd rather move all the properties to `Function` to avoid repetition and keep responsibilities clear (but that would break compatibility of course).

"Keeping responsibilities clear" is somehow ambiguous to me. There can be many things that would be natural to tie together in a class. I don't see why it is not the case here.

**Comment by niboshi** [comment]
Clearly `RawKernel` was originally intended to be a facility to compile a given CUDA source (#1398).
It was not introduced to be a `Function` wrapper.
The interface of `RawKernel.__init__` clearly reflects that intention.

Introducing a shortcut constructor that circumvent the primary role of the class, would make the role ambiguous and diverged from the original intention.

**Comment by niboshi** [comment]
If we were to redefine the role of `RawKernel` to be "a `Function` wrapper", there would be some inconsistencies.
* Interface of `RawKernel.__init__` would be erratic.
* The word `Raw` would not make sense. I guess it refers to a "*raw* CUDA source".
* How would it differ from `Function`?

**Comment by emcastillo** [comment]
IMO, I would keep `RawKernel` as a class whose main responsibility is to compile the kernel. Then I would just move all the CUDA attributes accessing code to the `Function` class. (Here it is possible to do a nocompat change, so the user would access the attributes through the `raw_kernel.function.num_regs` instead of the previous 
`raw_kernel.num_regs`. However, it is possible to keep the accessors in `RawKernel` as wrappers over `Function`

`RawModule.get_function` should now return `Function` instead of `RawKernels` and with the above change, compatibility is retained as the `Function` object is callable and all the properties can be retrieved.

**Comment by emcastillo** [comment]
I just pushed a preliminary version of my proposal, needs work (specifically with the cooperative groups as there is no easy way to propagate it directly to the function since it only gets instantiated in module, which doesn't feel natural to have such a parameter)

**Comment by niboshi** [comment]
@emcastillo Can you continue toward that direction?

**Comment by emcastillo** [comment]
I just submit a complete implementation.

Notes:
I deprecated the setter/getter methods in `RawKernel` and moved them to `Function`.
However my deprecation warnings or comments might not be compliant?

`enable_cooperative_groups` is now a parameter of `Module` as it is needed inside `Function`, but `Function` is never instantiated directly from `RawKernel` or `RawModule`. An alternative to this is to explicitly set it to `False` in the `Function.__init__` method and change the value from `RawKernel` or `RawModule` after getting the `Function` object.

**Comment by leofang** [comment]
> Notes:
I deprecated the setter/getter methods in RawKernel and moved them to Function.
However my deprecation warnings or comments might not be compliant?

If I may suggest, I would not deprecate setter/getter in `RawKernel`, and rather simply set to/get from the corresponding attribute of `Function`. Making them a thin wrapper, in other words. The reason is two-fold: 
1. we let `Function` float around in users' code, but we don't want them to notice this subtlety and would like them to think they are accessing `RawKernel`'s properties. 
2. `Function` is not documented and our life would be much simpler if it remains so, otherwise we will have to explain this in https://docs-cupy.chainer.org/en/stable/tutorial/kernel.html#raw-kernels.

**Comment by niboshi** [comment]
> 1. we let `Function` float around in users' code, but we don't want them to notice this subtlety and would like them to think they are accessing `RawKernel`'s properties.

The decision above (https://github.com/cupy/cupy/pull/2966#issuecomment-576513413) means we make `Function` public and documented. I think there's no problem letting users notice the `Function` class. 

> 2. `Function` is not documented and our life would be much simpler if it remains so, otherwise we will have to explain this in https://docs-cupy.chainer.org/en/stable/tutorial/kernel.html#raw-kernels, where the kernel attributes are documented.

Well, we will instead have to move the mention of the properties to `Function`'s documentation, if they're going to be deprecated in `RawKernel`.

**Comment by emcastillo** [comment]
I am not sure up to what extent this will affect user codes.

I am fine with both. I don't think that accessing execution time properties via function is a big deal.
But as you said we have to document it clearly.

**Comment by emcastillo** [comment]
Added a first draft for the docs.
As the `cupy.cuda.Function` class is not intended to be directly instantiated by users, it is a bit tricky to document.

Another concern, is that do we only expose `Function` for `RawKernel`? `ElementwiseKernel` uses `Function` too, but it does not save it in an attribute. I am not sure if we should unify this in all the kernel classes so there is some consistency.

**Comment by niboshi** [comment]
@emcastillo 

>Another concern, is that do we only expose `Function` for `RawKernel`? `ElementwiseKernel` uses `Function` too, but it does not save it in an attribute. I am not sure if we should unify this in all the kernel classes so there is some consistency.

It's not possible for `ElementwiseKernel` because `Function` can only be generated when `__call__` is called, which is not the case in `RawKernel`.

File: cupy/cuda/function.pyx [hunk 1]
```diff
@@ -155,7 +155,12 @@ cdef _launch(intptr_t func, Py_ssize_t grid0, int grid1, int grid2,
 
 cdef class Function:
 
-    """CUDA kernel function."""
+    """Class in charge of the execution of a compiled CUDA kernel function.
+
+    This class is used to access the runtime attributes of kernels compiled
+    via the :class:`~RawKernel`. It is mean to be accessed through
+    :attr:`~RawKernel.function`.
```
Comment (niboshi) [inline cupy/cuda/function.pyx]: This is not the primary goal of this class. I think we just don't need to mention this.

Comment (emcastillo) [reply cupy/cuda/function.pyx]: ok!

**Comment by emcastillo** [comment]
> It's not possible for ElementwiseKernel because Function can only be generated when __call__ is called, which is not the case in RawKernel.

Exactly, but it feels strange that for a `RawKernel` you can access those attributes and for an `ElementwiseKernel` you can't.

File: cupy/cuda/function.pyx [hunk 2]
```diff
@@ -155,7 +155,8 @@ cdef _launch(intptr_t func, Py_ssize_t grid0, int grid1, int grid2,
 
 cdef class Function:
 
-    """CUDA kernel function."""
+    """Class in charge of the execution of a compiled CUDA kernel function.
+    """
```
Comment (niboshi) [inline cupy/cuda/function.pyx]: Do you have to change the summary line?
Execution is not the only function of this class.
As we're adding kernel attributes, you may as well use this class to simply retrieve those attributes.

I think the previous expression represents exactly what this class is.

**Comment by niboshi** [comment]
>> It's not possible for ElementwiseKernel because Function can only be generated when __call__ is called, which is not the case in RawKernel.
>
>Exactly, but it feels strange that for a `RawKernel` you can access those attributes and for an `ElementwiseKernel` you can't. Anyway, it is not a problem of this PR.

I don't think it's strange, because it's the fundamental difference of those classes.
`RawKernel` can generate `Function` statically, and `ElementwiseKernel` generates it dynamically.
Simply that's reflected to the API.

**Comment by asi1024** [comment]
Could you resolve conflicts?

**Comment by emcastillo** [comment]
After the refactoring, the issue that this PR was addressing is gone, so we can close it.

**Pull request closed without merging**